testsuite/pytests/test_spike_train_injector.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+#
+# test_spike_train_injector.py
+#
+# This file is part of NEST.
+#
+# Copyright (C) 2004 The NEST Initiative
+#
+# NEST is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# NEST is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+import nest
+import pytest
+import numpy as np
+
+"""
+Test that spike_train_injector behaves as expected.
+
+This set of tests verifies that the spike_train_injector correctly handles
+the spike times for different options and emits spikes correctly in
+simulation.
+"""
+
+
+@pytest.fixture
+def reset_kernel():
+    nest.ResetKernel()
+
+
+@pytest.mark.parametrize(
+    ('in_spike_times', 'expected_spike_times', 'precise_times', 'allow_offgrid_times'),
+    [
+        ([1.0, 1.9999, 3.0001], [1.0, 2.0, 3.0], False, False),
+        ([1.0, 1.05, 3.0001],  [1.0, 1.1, 3.0], False, True),
+        ([1.0, 1.05, 3.0001],  [1.0, 1.05, 3.0001], True, False),
+    ]
+)
+def test_set_spike_times(
+    reset_kernel,
+    in_spike_times,
+    expected_spike_times,
+    precise_times,
+    allow_offgrid_times
+):
+    """
+    Verify correct handling of spike times not coinciding with a step for
+    different options. We set NEST to work with default resolution (step size)
+    of 0.1 ms and default tic length of 0.001 ms.
+    """
+    nest.set(resolution=0.1, tics_per_ms=1000.0)
+
+    inj_nrn = nest.Create("spike_train_injector",
+                          params={'spike_times': in_spike_times,
+                                  'precise_times': precise_times,
+                                  'allow_offgrid_times': allow_offgrid_times}
+                          )
+
+    out_spike_times = nest.GetStatus(inj_nrn, "spike_times")[0]
+    assert np.array_equal(out_spike_times, np.array(expected_spike_times))
```

Comment:
The neatest way I have found especially to get nice output if things go wrong is
```suggestion
    assert out_spike_times == approx(expected_spike_times)
```
This handles lists and arrays transparently. Needs `from pytest import approx`.